### PR TITLE
Add test cases for whitespace

### PR DESCRIPTION
**Pull request by SamMousa**
This test will catch implementation bugs related to whitespace.
https://github.com/jmespath/jmespath.js/issues/58

### CODE REVIEW

File: tests/basic.json [hunk 1]
```diff
@@ -18,6 +18,22 @@
...
             "result": "correct"
          },
+         {
+            "expression": "foo . bar . baz",
+            "result": "correct"
+         },
+         {
+            "expression": "foo\t.\tbar\t.\tbaz",
+            "result": "correct"
+         },
```
Comment (GrahamCampbell) [inline tests/basic.json]: Trailing whitespace here.

File: tests/basic.json [hunk 2]
```diff
@@ -18,6 +18,22 @@
...
+         {
+            "expression": "foo\t.\tbar\t.\tbaz",
+            "result": "correct"
+         },         
+         {
+            "expression": "foo\r.\rbar\r.\rbaz",
+            "result": "correct"
+         },
```
Comment (GrahamCampbell) [inline tests/basic.json]: and here

File: tests/basic.json [hunk 3]
```diff
@@ -18,6 +18,22 @@
...
+            "result": "correct"
+         },         
+         {
+            "expression": "foo\r.\rbar\r.\rbaz",
+            "result": "correct"
+         },      
+         {
+            "expression": "foo\r\n.\r\nbar\r\n.\r\nbaz",
+            "result": "correct"
+         },
```
Comment (GrahamCampbell) [inline tests/basic.json]: and here

File: tests/basic.json (hunk 1 above)
Comment (SamMousa) [reply tests/basic.json]: Are you talking about the code or the test case?
Do you want the test expression to have a trailing `\t`?

Comment (GrahamCampbell) [reply tests/basic.json]: No, in the actual json document. ;)

image

Comment (GrahamCampbell) [reply tests/basic.json]: Should have said: "Please can you remove the trailing space characters on this line". Sorry!

Comment (SamMousa) [reply tests/basic.json]: The one time I don't use my own IDE, 20% of LOC contain errors =D

**Review by GrahamCampbell — APPROVED** [review]
Looks good. My comments seem to have not hidden themselves, but they are no longer valid.

**Comment by jamesls** [comment]
I know this is super, super late for feedback, but I won't be able to take this change.  While most of the implementations ignore some degree of whitespace at the lexing stage, the [grammar](https://jmespath.org/specification.html#grammar) for jmespath uses ABNF which is strict about requiring whitespace to be explicitly modeled.  From [Section 3.1 of RFC 4234](https://datatracker.ietf.org/doc/html/rfc4234#section-3.1):

> NOTE:
>
>    This specification for ABNF does not provide for implicit
>    specification of linear white space.
>
> Any grammar that wishes to permit linear white space around
> delimiters or string segments must specify it explicitly.  It is
> often useful to provide for such white space in "core" rules that are
> then used variously among higher-level rules.  The "core" rules might
> be formed into a lexical analyzer or simply be part of the main
> ruleset.

If we're going to support whitespace around expressions we would need to update the spec/grammar first, and I'm not sure at this point it's something we'd want to do (though I'm open to a discussion about it).
